### PR TITLE
Support cleaning up expired transactions by number

### LINKED ISSUE #2410: Support cleaning up expired transactions by number
Currently, FE can only clean load transactions at the specified interval, default three days.
The memory usage is relatively large if the load frequency is high, triggering FE Full GC.

截屏2021-12-24 下午1 33 07
截屏2021-12-24 下午1 33 45

Solution:
1. Support cleaning up expired transactions by number.
2. For non-expired transactions, clear publish version tasks when the transaction state change to be visible.

### PR DESCRIPTION
**Pull request by wyb**
Currently, FE can only clean load transactions at the specified interval, default three days.
The memory usage is relatively large if the load frequency is high, triggering FE Full GC.
The pull request adds two strategies to reclaim the memory.
1. Support cleaning up expired transactions by number.
2. For non-expired transactions, clear publish version tasks when the transaction state change to be visible.

fix #2410

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/transaction/PublishVersionDaemon.java [hunk 1]
```diff
@@ -119,15 +119,14 @@ private void publishVersion() throws UserException {
...
-                    // if finish transaction state failed, then update publish version time, should check 
-                    // to finish after some interval
                     transactionState.updateSendTaskTime();
                     LOG.debug("publish version for transation {} failed, has {} error replicas during publish",
                             transactionState, publishErrorReplicaIds.size());
                 } else {
                     for (PublishVersionTask task : transactionState.getPublishVersionTasks().values()) {
                         AgentTaskQueue.removeTask(task.getBackendId(), TTaskType.PUBLISH_VERSION, task.getSignature());
                     }
+                    transactionState.clearPublishVersionTasks();
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/transaction/PublishVersionDaemon.java:129]: Please provide some comments here to explain why calling this tasks.

File: fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java [hunk 2]
```diff
@@ -1122,17 +1122,18 @@ private void clearBackendTransactions(TransactionState transactionState) {
     public void removeExpiredTxns(long currentMillis) {
         writeLock();
         try {
+            int numJobsToRemove = getTransactionNum() - Config.label_keep_max_num;
             while (!finalStatusTransactionStateDeque.isEmpty()) {
                 TransactionState transactionState = finalStatusTransactionStateDeque.getFirst();
-                if (transactionState.isExpired(currentMillis)) {
+                if (transactionState.isExpired(currentMillis) || numJobsToRemove > 0) {
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java:1128]: what is the finalStatusTransactionStateDeque sorted by?
I think we'd better to keep the latest transactions.

Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/transaction/DatabaseTransactionMgr.java:1128]: By finish time.

File: fe/fe-core/src/main/java/com/starrocks/transaction/PublishVersionDaemon.java (hunk 1 above)
Comment (wyb) [reply fe/fe-core/src/main/java/com/starrocks/transaction/PublishVersionDaemon.java:130]: ok

**Review by Astralidea — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**